pkg/_internal/gracePeriod/gracePeriod.go

```diff
@@ -0,0 +1,57 @@
+package gracePeriod
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/Kuadrant/multicluster-gateway-controller/pkg/_internal/metadata"
+)
+
+const (
+	graceTimestampAnnotation = "kuadrant.io/grace-timeout"
+)
+
+var ErrGracePeriodNotExpired = fmt.Errorf("grace period has not yet expired")
+
+func WasGracePeriodNotExpiredErr(e error) bool {
+	return strings.Contains(e.Error(), "grace period has not yet expired")
+}
+
+func GracefulDelete(ctx context.Context, c client.Client, obj client.Object, at time.Time) error {
+	if err := c.Get(ctx, client.ObjectKeyFromObject(obj), obj); err != nil {
+		return err
+	}
+
+	// if At is before the current time, just delete it
+	if at.Before(time.Now()) || at.Equal(time.Now()) {
+		return c.Delete(ctx, obj)
+	}
```

Comment:
so I am not getting why we need more than this?

Reply:
the `at` timestamp is recalculated on each reconcile, this is simply handling for if the grace period is zero and no TTL was found, then there's no need to fuss around with annotations, just delete it.

But the annotation below is required if the `at` is anything later than now, so we know when to remove it, the new at would only be honoured if the previous at annotation was somehow corrupted or removed manually.